internal/impl/kubernetes/input_watch.go

```diff
@@ -0,0 +1,506 @@
+	if err != nil {
+		return nil, err
+	}
+	k.labelSelector = LabelSelectorFromMap(labelSelectorMap)
+	fieldSelectorMap, err := conf.FieldStringMap("field_selector")
+	if err != nil {
+		return nil, err
+	}
+	k.fieldSelector = LabelSelectorFromMap(fieldSelectorMap)
+
+	// Parse event types filter
+	eventTypesList, err := conf.FieldStringList("event_types")
+	if err != nil {
+		return nil, err
+	}
+	k.eventTypes = make(map[string]struct{})
+	for _, et := range eventTypesList {
+		k.eventTypes[et] = struct{}{}
+	}
+
+	// Parse behavior options
+	if k.includeInitialList, err = conf.FieldBool("include_initial_list"); err != nil {
+		return nil, err
+	}
+	requestTimeoutStr, err := conf.FieldString("request_timeout")
+	if err != nil {
+		return nil, err
+	}
+	if k.requestTimeout, err = time.ParseDuration(requestTimeoutStr); err != nil {
+		return nil, fmt.Errorf("failed to parse request_timeout: %w", err)
+	}
+
+	// Parse backoff configuration
+	if k.backoffCtor, err = retries.CommonRetryBackOffCtorFromParsed(conf); err != nil {
+		return nil, fmt.Errorf("failed to parse backoff config: %w", err)
+	}
+
+	// Get Kubernetes client (needed for RESTMapper)
+	if k.clientSet, err = GetClientSet(context.Background(), conf, mgr.FS()); err != nil {
+		return nil, fmt.Errorf("failed to create kubernetes client: %w", err)
+	}
+
+	// Determine GVR from resource or custom_resource
+	resource, _ := conf.FieldString("resource")
+	if resource != "" {
+		// Use RESTMapper to dynamically resolve the GVR
+		gvr, err := k.clientSet.Mapper.ResourceFor(schema.GroupVersionResource{Resource: resource})
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve resource %q: %w", resource, err)
+		}
+		k.gvr = gvr
+	} else {
+		// Custom resource with explicit GVR
+		crConf := conf.Namespace("custom_resource")
+		group, _ := crConf.FieldString("group")
+		version, _ := crConf.FieldString("version")
+		crResource, _ := crConf.FieldString("resource")
+
+		if crResource == "" {
+			return nil, errors.New("custom_resource.resource is required when using custom_resource")
+		}
+
+		k.gvr = schema.GroupVersionResource{
+			Group:    group,
+			Version:  version,
+			Resource: crResource,
+		}
+	}
+
+	return k, nil
+}
+
+func (k *kubernetesWatchInput) getResourceVersion(namespace string) string {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+	return k.resourceVers[namespace]
+}
+
+func (k *kubernetesWatchInput) setResourceVersion(namespace, version string) {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+	if version == "" {
+		delete(k.resourceVers, namespace)
+		return
+	}
+	k.resourceVers[namespace] = version
+}
+
+func (k *kubernetesWatchInput) Connect(ctx context.Context) error {
+	// Start watch loops for each namespace
+	namespaces := k.namespaces
+	if len(namespaces) == 0 {
+		namespaces = []string{""} // Empty string = all namespaces
+	}
+
+	for _, ns := range namespaces {
+		k.wg.Add(1)
+		go func(namespace string) {
+			defer k.wg.Done()
+			k.watchNamespace(namespace)
+		}(ns)
+	}
```

Comment:
We can use the fancy new waitgroup method 👇 

```suggestion
	for _, ns := range namespaces {
		k.wg.Go(func() {
			k.watchNamespace(ns)
		})
	}
```